File: Emulator/src/main/core/net/event/GameWriteEvent.java
package core.net.event;

import core.net.IoSession;
import core.net.IoWriteEvent;
import core.net.packet.IoBuffer;

import java.nio.ByteBuffer;

/**
 * The type Game write event.
 */
public final class GameWriteEvent extends IoWriteEvent {

    /**
     * Instantiates a new Game write event.
     *
     * @param session the session
     * @param context the context
     */
    public GameWriteEvent(IoSession session, Object context) {
        super(session, context);
    }

    @Override
    public void write(IoSession session, Object context) {
        if (context instanceof ByteBuffer) {
            session.queue((ByteBuffer) context);
            return;
        }
        IoBuffer buffer = (IoBuffer) context;
        ByteBuffer buf = (ByteBuffer) buffer.toByteBuffer().flip();
        if (buf == null) {
            throw new RuntimeException("Critical networking error: The byte buffer requested was null.");
        }
        if (buffer.opcode() != -1) {
            int packetLength = buf.remaining() + 4;
            ByteBuffer response = ByteBuffer.allocate(packetLength);
            response.put((byte) buffer.opcode());
            switch (buffer.getHeader()) {
                case BYTE:
                    response.put((byte) buf.remaining());
                    break;
                case SHORT:
                    response.putShort((short) buf.remaining());
                    break;
                default:
                    break;
            }
            buf = (ByteBuffer) response.put(buf).flip();
        }
        session.queue(buf);
    }

}

File: Emulator/src/main/core/net/event/GameWriteEvent.kt
package core.net.event

import core.net.IoSession
import core.net.IoWriteEvent
import core.net.packet.IoBuffer
import core.net.packet.PacketHeader
import java.nio.ByteBuffer

/**
 * Handles game packet writing events.
 *
 * @author Emperor
 */
class GameWriteEvent(session: IoSession, context: Any) : IoWriteEvent(session, context) {

    override fun write(session: IoSession, context: Any) {
        when (context) {
            is ByteBuffer -> {
                session.queue(context)
                return
            }
            is IoBuffer -> {
                var buf = context.toByteBuffer().flip() as ByteBuffer?
                    ?: throw RuntimeException("Critical networking error: The byte buffer requested was null.")

                if (context.opcode() != -1) {
                    val packetLength = buf.remaining() + 4
                    val response = ByteBuffer.allocate(packetLength)
                    response.put(context.opcode().toByte())

                    when (context.header) {
                        PacketHeader.BYTE -> response.put(buf.remaining().toByte())
                        PacketHeader.SHORT -> response.putShort(buf.remaining().toShort())
                        else -> { /* do nothing */ }
                    }

                    response.put(buf)
                    response.flip()
                    buf = response
                }
                session.queue(buf)
            }
            else -> throw IllegalArgumentException("Unsupported context type: ${context::class}")
        }
    }
}
